cli/azd/pkg/ai/config.go

```diff
@@ -0,0 +1,80 @@
+package ai
+
+import (
+	"fmt"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+	"gopkg.in/yaml.v3"
+)
+
+// ComponentConfig is a base configuration structure used by multiple AI components
+type ComponentConfig struct {
+	Name      osutil.ExpandableString            `yaml:"name,omitempty"`
+	Path      string                             `yaml:"path,omitempty"`
+	Overrides map[string]osutil.ExpandableString `yaml:"overrides,omitempty"`
+}
+
+// EndpointDeploymentConfig is a configuration structure for an ML online endpoint deployment
+type EndpointDeploymentConfig struct {
+	Workspace   osutil.ExpandableString `yaml:"workspace,omitempty"`
+	Environment *ComponentConfig        `yaml:"environment,omitempty"`
+	Model       *ComponentConfig        `yaml:"model,omitempty"`
+	Flow        *ComponentConfig        `yaml:"flow,omitempty"`
+	Deployment  *ComponentConfig        `yaml:"deployment,omitempty"`
+}
+
+// Flow is a configuration to defined a Prompt flow component
+type Flow struct {
+	Name        string            `json:"name"`
+	Description string            `json:"description"`
+	Type        string            `json:"type"`
+	Path        string            `json:"path"`
+	DisplayName string            `json:"display_name"`
+	Tags        map[string]string `json:"tags"`
+}
+
+// Scope is a context based structure to define the Azure scope of a AI component
+type Scope struct {
+	subscriptionId string
+	resourceGroup  string
+	workspace      string
+}
```

Comment:
The goal was to make this `Scope` immutable so that after you create it you cannot inadvertently change the values.

Reply:
Can we stay away from this immutable pattern?  It is not natural in go.

Go makes it easy to copy an entire struct with a simple assignment, so we may think that passing arguments by value is all that is needed to have immutability at the cost of copying.

For example
```go
s1 := Scope{SubscriptionId: "a"}
callSomeFuntction(s1) // s1 is copied, so it can't be changed!!!  
```

Coming from Java/C#, I understand the concept/approach you are trying to build here, but I think azd should feel more go than any other language.

Reply:
To me this doesn't feel like any different language - it is just a concept.  This `Scope` requires a subscriptionId, resource group and workspace name.  It is initialized and passed into functions.  The scope is not intended to be modified after its creation so therefore does not export the field directly and instead opts to only expose read-only getters for that data.

Reply:
>  it is just a concept.

Concepts are enclosed within paradigms. In this case, `encapsulation` is a concept introduced by the Object Oriented paradigm.
As every paradigm, the reason behind the concepts are based on the idea of `that's how it works, how it has been and the best known way how it has to be`.  

Languages like Java and C# took OOP very seriously by natively supporting every concept of the OOP (classes, encapsulation, inheritance, overloading, etc). 
But Go combines elements from procedural, object-oriented, and concurrent programming paradigms. It emphasizes simplicity, efficiency, and safety as the main pillars to have in mind.  In the go world, encapsulation goes against simplicity.

Reply:
> It is initialized and passed into functions

When the `Scope` is passed to functions, a copy is created.  Any changes to the `Scope` within the function will not mutate the `Scope` outside of the function.  That's the go-way to simplify immutability.
If you come from a C++ world, you could prefer passing a `*Scope` to avoid making copies of the strings in the `Scope`, but go-compiler has proven to be really fast optimizing those copies, doesn't really impacting performance.

So, you just need to take `Scope` as argument in the functions, and be sure that no one will be able to change the values from that scope inside the function.

Reply:
I think the value is this design is that it is clear that after the object is created you are not expected to change the value.  The fact that it is passed "by value" or "by reference" feels orthogonal to me.  The intention is that this is a single value that once created and you can address parts of but you can't change parts of.

I think of this as information hiding, we don't want other parts of the system to take a dependency on the internal state that makes up the `Scope` object, you should use the accessors, and allow the implementation of the Scope type to control how it stores the state.